Server doesn't stop succesfully in CLI after opening WebSocket connections

<!-- 🚀 Thank you for contributing! --->

<!-- Provide a brief description of the issue here. -->
<!-- Pretend you are explaining it to a friend, not yourself! -->

### Steps to reproduce

Create a simple Vapor app that opens WebSocket connections:

```swift
import Vapor

extension Application {
  func configure(
    onWebSocketOpen: @escaping (WebSocket) -> (),
    onWebSocketClose: @escaping (WebSocket) -> ()
  ) {
    webSocket("watcher") { _, ws in
      onWebSocketOpen(ws)
      ws.onClose.whenComplete { _ in onWebSocketClose(ws) }
    }
  }
}

final class Server {
  private struct Lifecycle: LifecycleHandler {
    weak var server: Server?

    /// Closes all active WebSocket connections
    func shutdown(_ app: Application) {
      try! EventLoopFuture<()>.andAllSucceed(
        server?.connections.map { $0.close() } ?? [],
        on: app.eventLoopGroup.next()
      ).wait()
      print("shutdown finished")
    }
  }

  private var connections = Set<WebSocket>()
  private let app: Application

  init() throws {
    var env = Environment(name: "development", arguments: ["vapor"])
    try LoggingSystem.bootstrap(from: &env)
    app = Application(env)
    app.configure(
      onWebSocketOpen: { [weak self] in
        self?.connections.insert($0)
      },
      onWebSocketClose: { [weak self] in
        self?.connections.remove($0)
      }
    )
  }

  /// Blocking function that starts the HTTP server
  func run() throws {
    defer { app.shutdown() }
    app.lifecycle.use(Lifecycle(server: self))
    try app.run()
  }
}

try Server().run()
```

### Expected behavior

When this server app is started from command-line and any WebSocket connections are established, the server process can stopped quickly with Ctrl-C.

### Actual behavior

The server process can't be stopped quickly with Ctrl-C. It hangs for about 5 seconds or so and then raises the following error:

```
shutdown finished
[ ERROR ] Could not stop HTTP server: Abort.500: Server stop took too long.
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
```

### Environment

These are the details of revelant package dependencies from `Package.resolved`:

```
      {
        "package": "vapor",
        "repositoryURL": "https://github.com/vapor/vapor.git",
        "state": {
          "branch": null,
          "revision": "88293674e2ea017691c56af20d0938dfff7ece04",
          "version": "4.27.0"
        }
      },
      {
        "package": "websocket-kit",
        "repositoryURL": "https://github.com/vapor/websocket-kit.git",
        "state": {
          "branch": null,
          "revision": "b0736014be634475dac4c23843811257d86dcdc1",
          "version": "2.1.1"
        }
      }
```

* Vapor Framework version: 4.27.0, but this was reproducible in versions as old as 4.5.0
* Vapor Toolbox version: not installed
* OS version: macOS 10.15.6

@tanner0101 I did some further digging into this and here is what I was able to come up with:

- The issue appears to be sporadic on macOS. There are times when terminating an application from Xcode exits with code 9 and rebuilding/running it within Xcode continues to work fine. There are other times when attempting to build/run the project a SECOND time exits with code -1, along with an error message indicating that the address (default: 127.0.0.1:8080) is in use. Attached is the `Package.resolved` file for the project I used to test this on macOS (version: 10.15.6 including the latest supplemental update and Xcode version: 11.6).

- When the issue does occur on macOS, the server continues to be responsive even though the project is terminated on Xcode. A quick check on Activity Monitor shows me a program called Run that must be terminated using Force Quit in order to terminate the server and be able to successfully build the project again.

- The issue is consistent on Linux. The bug appears every single time I attempt it on any Ubuntu release (16.04, 18.04, 20.04).

- The issue continues to persist with the latest release of Vapor 4.27.3 on both macOS (sporadically) and Linux (consistently).

- Ubuntu 20.04 was able to give me a little more info to indicate where the problem might be. Below are the error messages received (the last error was repeated twice every time):

```
[ERROR] Could not stop HTTP server: abort.500: Server stop took too long.
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
ERROR: Cannot schedule tasks on an EventLoop that has already shut down. This will be upgraded to a forced crash in future SwiftNIO versions.
```

--- a/Sources/Vapor/HTTP/Headers/HTTPHeaders+Connection.swift
+++ b/Sources/Vapor/HTTP/Headers/HTTPHeaders+Connection.swift
@@ -1,0 +1,29 @@
+extension HTTPHeaders {
+    public struct Connection: ExpressibleByStringLiteral, Equatable {
+        public static let close: Self = "close"
+        public static let keepAlive: Self = "keep-alive"
+
+        public let value: String
+
+        public init(value: String) {
+            self.value = value
+        }
+
+        public init(stringLiteral value: String) {
+            self.init(value: value)
+        }
+    }
+
+    public var connection: Connection? {
+        get {
+            self.first(name: .connection).flatMap(Connection.init(value:))
+        }
+        set {
+            if let value = newValue {
+                self.replaceOrAdd(name: .connection, value: value.value)
+            } else {
+                self.remove(name: .connection)
+            }
+        }
+    }
+}

--- a/Sources/Vapor/HTTP/Server/HTTPServer.swift
+++ b/Sources/Vapor/HTTP/Server/HTTPServer.swift
@@ -361,7 +361,7 @@ extension ChannelPipeline {
         handlers.append(serverResEncoder)
         
         // add server request -> response delegate
-        let handler = HTTPServerHandler(responder: responder)
+        let handler = HTTPServerHandler(responder: responder, logger: application.logger)
         handlers.append(handler)
         
         return self.addHandlers(handlers).flatMap {
@@ -426,7 +426,7 @@ extension ChannelPipeline {
         )
         handlers.append(serverReqDecoder)
         // add server request -> response delegate
-        let handler = HTTPServerHandler(responder: responder)
+        let handler = HTTPServerHandler(responder: responder, logger: application.logger)
 
         // add HTTP upgrade handler
         let upgrader = HTTPServerUpgradeHandler(

--- a/Sources/Vapor/HTTP/Server/HTTPServerHandler.swift
+++ b/Sources/Vapor/HTTP/Server/HTTPServerHandler.swift
@@ -5,9 +5,13 @@ final class HTTPServerHandler: ChannelInboundHandler, RemovableChannelHandler {
     typealias OutboundOut = Response
     
     let responder: Responder
+    let logger: Logger
+    var isShuttingDown: Bool
     
-    init(responder: Responder) {
+    init(responder: Responder, logger: Logger) {
         self.responder = responder
+        self.logger = logger
+        self.isShuttingDown = false
     }
     
     func channelRead(context: ChannelHandlerContext, data: NIOAny) {
@@ -34,18 +38,32 @@ final class HTTPServerHandler: ChannelInboundHandler, RemovableChannelHandler {
         case 2:
             context.write(self.wrapOutboundOut(response), promise: nil)
         default:
-            response.headers.add(name: .connection, value: request.isKeepAlive ? "keep-alive" : "close")
+            let keepAlive = !self.isShuttingDown && request.isKeepAlive
+            if self.isShuttingDown {
+                self.logger.debug("In-flight request has completed")
+            }
+            response.headers.add(name: .connection, value: keepAlive ? "keep-alive" : "close")
             let done = context.write(self.wrapOutboundOut(response))
             done.whenComplete { result in
                 switch result {
                 case .success:
-                    if !request.isKeepAlive {
+                    if !keepAlive {
                         context.close(mode: .output, promise: nil)
                     }
                 case .failure(let error):
                     self.errorCaught(context: context, error: error)
                 }
             }
+        }
+    }
+
+    func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+        switch event {
+        case is ChannelShouldQuiesceEvent:
+            self.logger.trace("HTTP handler will no longer respect keep-alive")
+            self.isShuttingDown = true
+        default:
+            self.logger.trace("Unhandled user event: \(event)")
         }
     }
 }

--- a/Sources/Vapor/HTTP/Server/HTTPServerRequestDecoder.swift
+++ b/Sources/Vapor/HTTP/Server/HTTPServerRequestDecoder.swift
@@ -17,13 +17,14 @@ final class HTTPServerRequestDecoder: ChannelInboundHandler, RemovableChannelHan
     var requestState: RequestState
     var bodyStreamState: HTTPBodyStreamState
 
-    private let logger: Logger
+    var logger: Logger {
+        self.application.logger
+    }
     var application: Application
     
     init(application: Application) {
         self.application = application
         self.requestState = .ready
-        self.logger = Logger(label: "codes.vapor.server")
         self.bodyStreamState = .init()
     }
     
@@ -179,7 +180,8 @@ final class HTTPServerRequestDecoder: ChannelInboundHandler, RemovableChannelHan
     }
 
     func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
-        if event is HTTPServerResponseEncoder.ResponseEndSentEvent {
+        switch event {
+        case is HTTPServerResponseEncoder.ResponseEndSentEvent:
             switch self.requestState {
             case .streamingBody(let bodyStream):
                 // Response ended during request stream.
@@ -197,6 +199,17 @@ final class HTTPServerRequestDecoder: ChannelInboundHandler, RemovableChannelHan
                 // Response ended after request had been read.
                 break
             }
+        case is ChannelShouldQuiesceEvent:
+            switch self.requestState {
+            case .ready:
+                self.logger.trace("Closing keep-alive HTTP connection since server is going away")
+                context.channel.close(mode: .all, promise: nil)
+            default:
+                self.logger.debug("A request is currently in-flight")
+                context.fireUserInboundEventTriggered(event)
+            }
+        default:
+            self.logger.trace("Unhandled user event: \(event)")
         }
     }
 }

--- a/Tests/VaporTests/ServerTests.swift
+++ b/Tests/VaporTests/ServerTests.swift
@@ -277,7 +277,7 @@ final class ServerTests: XCTestCase {
             request: request,
             delegate: response
         ).wait()
-
+        
         XCTAssertEqual(context.server, ["foo", "bar", "baz"])
         XCTAssertEqual(context.client, ["foo", "bar", "baz"])
     }
@@ -313,6 +313,27 @@ final class ServerTests: XCTestCase {
         XCTAssertEqual(a.status, .ok)
         let b = try app.http.client.shared.execute(request: request).wait()
         XCTAssertEqual(b.status, .ok)
+    }
+
+    func testQuiesceKeepAliveConnections() throws {
+        let app = Application(.testing)
+        defer { app.shutdown() }
+
+        app.get("hello") { req in
+            "world"
+        }
+
+        let port = 1337
+        app.http.server.configuration.port = port
+        try app.start()
+
+        let request = try HTTPClient.Request(
+            url: "http://localhost:\(port)/hello",
+            method: .GET,
+            headers: ["connection": "keep-alive"]
+        )
+        let a = try app.http.client.shared.execute(request: request).wait()
+        XCTAssertEqual(a.headers.connection, .keepAlive)
     }
 
     override class func setUp() {

--- a/Tests/VaporTests/WebSocketTests.swift
+++ b/Tests/VaporTests/WebSocketTests.swift
@@ -76,4 +76,84 @@ final class WebSocketTests: XCTestCase {
 
         try XCTAssertEqual(promise.futureResult.wait(), "foo")
     }
+
+    func testLifecycleShutdown() throws {
+        let app = Application(.testing)
+        app.http.server.configuration.port = 1337
+
+        final class WebSocketManager: LifecycleHandler {
+            private let lock: Lock
+            private var connections: Set<WebSocket>
+
+            init() {
+                self.lock = .init()
+                self.connections = .init()
+            }
+
+            func track(_ ws: WebSocket) {
+                self.lock.lock()
+                defer { self.lock.unlock() }
+                self.connections.insert(ws)
+                ws.onClose.whenComplete { _ in
+                    self.lock.lock()
+                    defer { self.lock.unlock() }
+                    self.connections.remove(ws)
+                }
+            }
+
+            func broadcast(_ message: String) {
+                self.lock.lock()
+                defer { self.lock.unlock() }
+                for ws in self.connections {
+                    ws.send(message)
+                }
+            }
+
+            /// Closes all active WebSocket connections
+            func shutdown(_ app: Application) {
+                self.lock.lock()
+                defer { self.lock.unlock() }
+                app.logger.debug("Shutting down \(self.connections.count) WebSocket(s)")
+                try! EventLoopFuture<Void>.andAllSucceed(
+                    self.connections.map { $0.close() } ,
+                    on: app.eventLoopGroup.next()
+                ).wait()
+            }
+        }
+
+        let webSockets = WebSocketManager()
+        app.lifecycle.use(webSockets)
+
+        app.webSocket("watcher") { req, ws in
+            webSockets.track(ws)
+            ws.send("hello")
+        }
+
+        try app.start()
+
+        let clientGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+        defer { try! clientGroup.syncShutdownGracefully() }
+        let connectPromise = app.eventLoopGroup.next().makePromise(of: WebSocket.self)
+        WebSocket.connect(to: "ws://localhost:1337/watcher", on: clientGroup) { ws in
+            connectPromise.succeed(ws)
+        }.cascadeFailure(to: connectPromise)
+
+        let ws = try connectPromise.futureResult.wait()
+        app.shutdown()
+        try ws.onClose.wait()
+    }
+
+    override class func setUp() {
+        XCTAssertTrue(isLoggingConfigured)
+    }
+}
+
+extension WebSocket: Hashable {
+    public static func == (lhs: WebSocket, rhs: WebSocket) -> Bool {
+        lhs === rhs
+    }
+
+    public func hash(into hasher: inout Hasher) {
+        ObjectIdentifier(self).hash(into: &hasher)
+    }
 }

